include/xtensor/xreducer.hpp

```diff
@@ -37,20 +37,231 @@ namespace xt
 #endif
 
+    template <class F, class E, class X>
+    auto reduce_immediate(F&& f, E&& e, X&& axes)
+    {
+        using shape_type = std::vector<std::size_t>;
+        using accumulate_functor = std::decay_t<decltype(std::get<0>(f))>;
+        using result_type = typename accumulate_functor::result_type;
+
+        // retrieve functors from triple struct
+        auto acc_fct = std::get<0>(f);
+        auto init_fct = std::get<1>(f);
+        auto merge_fct = std::get<2>(f);
+
+        shape_type result_shape(e.dimension() - axes.size());
+        std::vector<std::size_t> iter_shape = e.shape();
+        std::vector<std::size_t> iter_strides(e.dimension());
+
+        xt::xarray<result_type> result;
+
+        if (!std::is_sorted(axes.cbegin(), axes.cend()))
+        {
+            throw std::runtime_error("Reducing axes should be sorted");
+        }
+
+        if (e.dimension() == axes.size())
+        {
+            auto begin = e.data().begin();
+            result_type tmp = init_fct(*begin);
+            ++begin;
+            result(0) = std::accumulate(begin, e.data().end(), tmp, acc_fct);
+            return result;
+        }
+
+        std::size_t idx = 0;
+        for (std::size_t i = 0; i < e.dimension(); ++i)
+        {
+            if (std::find(axes.begin(), axes.end(), i) == axes.end())
+            {
+                // i not in axes!
+                result_shape[idx] = e.shape()[i];
+                ++idx;
+            }
+        }
+
+        result.reshape(result_shape);
+
+        std::size_t inner_loop_size = e.strides()[axes[axes.size() - 1]];
+        std::size_t inner_stride    = e.strides()[axes[axes.size() - 1]];
+        std::size_t outer_loop_size = e.shape()[axes[axes.size() - 1]];
+
+        auto it = axes.rbegin();
+        auto last_ax = *it;
+        ++it;
+        for (; it != axes.rend(); ++it)
+        {
+            if (*it == last_ax - 1)
+            {
+                last_ax = *it;
+                outer_loop_size *= e.shape()[last_ax];
+            }
+        }
+
+        idx = 0;
+        for (std::size_t i = 0; i < e.dimension(); ++i)
+        {
+            if (std::find(axes.begin(), axes.end(), i) == axes.end())
+            {
+                // i not in axes!
+                iter_strides[i] = result.strides()[idx];
+                ++idx;
+            }
+        }
+
+        // TODO remove all entries where iter_shape == 0
+        iter_shape.erase(iter_shape.begin() + last_ax, iter_shape.end());
+        iter_strides.erase(iter_strides.begin() + last_ax, iter_strides.end());
+
+        for (std::size_t i = 0; i < iter_shape.size(); ++i)
+        {
+            if (iter_shape[i] == 0)
+            {
+                iter_shape.erase(iter_shape.begin() + i);
+                iter_strides.erase(iter_strides.begin() + i);
+                ++i;
+            }
+        }
+
+        xindex temp_idx(iter_shape.size());
+        auto next_idx = [&iter_shape, &iter_strides, &temp_idx]()
```

Comment:
why capturing `temp_idx` instead of passing it as const reference ?

Reply:
temp_idx holds the state of the iteration and is modified in the loop. you mean instead as a function argument?

Reply:
yep sorry I mean as an argument of the next_id function. And indeed it should be a simple reference, not a const reference. It's just that I find it clearer to have (but that's totally subjective)
```cpp
auto res = next_idx(temp_idx);
```

Reply:
the thing is that temp_idx is basically internal to the functor/lambda here, and isn't used outside. that's why i decided to do it like this